internal/wasm/jit/jit_arm64.go

```diff
@@ -69,14 +70,26 @@ const (
 // engine is the pointer to the "*engine" as uintptr.
 func jitcall(codeSegment, engine uintptr)
 
+// golang-asm is not goroutine-safe so we take lock until we complete the compilation.
+// TODO: delete after https://github.com/tetratelabs/wazero/issues/233
+var assemblerMutex = &sync.Mutex{}
+
+func unlockAssembler() {
+	assemblerMutex.Unlock()
+}
+
 // newCompiler returns a new compiler interface which can be used to compile the given function instance.
 // Note: ir param can be nil for host functions.
-func newCompiler(f *wasm.FunctionInstance, ir *wazeroir.CompilationResult) (compiler, error) {
+func newCompiler(f *wasm.FunctionInstance, ir *wazeroir.CompilationResult) (c compiler, done func(), err error) {
+	// golang-asm is not goroutine-safe so we take lock until we complete the compilation.
+	// TODO: delete after https://github.com/tetratelabs/wazero/issues/233
+	assemblerMutex.Lock()
```

Comment:
I may be missing some contexts, but what is the reason `newCompiler` is returning the unlock function and asking the caller to call it, instead of `newCompiler` itself calling `defer assemblerMutex.Unlock()` or something like that?

Reply:
oops never mind. I've realized that the caller is not calling `done` but deferring it, and this `done` must be called after all compile related functions complete